### DISCUSSION TITLE
Pinout for 5-pin SerialUPDI?

### DISCUSSION THREAD

**Discussion by jvasileff** [SpenceKonde/megaTinyCore#603] [Q&A]
I've looked around, but haven't seen any existing 5-pin pinouts for SerialUPDI, which would include TX and RX.

I'm wondering what would be best for a programmer that would look something like:

<img width="671" alt="Screen Shot 2022-01-10 at 1 05 36 PM" src="https://user-images.githubusercontent.com/236050/148816312-c1cd17dd-7058-4f53-94ba-b10f9f9534ce.png">

The 3-pin [SerialUPDI](https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md) pinout is:

1. UPDI
2. GND
3. VCC

My current idea for 5-pin programming+serial connector is:

1. UPDI
2. GND
3. VCC
4. TX
5. RX

The first three pins would match the 3-pin connector for compatibility. With TX and RX occupying pin 4 and 5, I *think* reverse connection would be *somewhat* safe:

Target MCU Pin | Reversed Programmer Pin
---- | ----
UPDI (PIN_A1) | RX Hi-Z, maybe weak pullup
GND | TX via 1k resistor (could be high or low)
VCC | VCC
RX (PIN_PB3) | GND
TX (PIN_PB2) | UPDI out via > 1k resistor (could be high or low)

VCC would always be connected correctly; reverse voltage would never be applied. With the connection reversed, the target would be fully powered if grounded by some other connection.

If the target is not otherwise grounded, it will be grounded through the RX/PIN_PB3's protection diodes. This isn't good, but likely won't immediately destroy the chip.

If the target is otherwise grounded, and RX/PIN_PB3 is set to output HIGH, there will be a direct short through PIN_PB3. But... it would be a mistake anyway for RX to be set as an output.

Am I missing anything? Any other/better ideas?

**Comment by SpenceKonde** [comment]
UPDI and Vcc can be safely reversed, because the UPDI pin is HV tolerant. (up to 12v for HV UPDI) . 

If using a polarized plug, reverse connections are much less of a concern (as shown in that picture - you do need two serial adapters or some sort of switch to do programming and UPDI with a single cable, though - so a single cable is somewhat less appealing than it otherwise would be. 

I'd use the FTDI pinout with the pin that's nominally CTS connected to the UPDI pin (in fact, that is what I  *do* do on the latest revisions of my breakout boards). It is reasonably safe when reversed. If there is no external grounding, because ground is on the DTR line, the other end of which, if connected to anything, is connected to a 0.1uf cap for autoreset (almost always done on the x26 and x27 parts), so the rest of the board won't be grounded at all. And even if it is, you're okay as long as the adapter's TX pin is connected through anything resistor like; see absolute maximums section of datasheet, maximum pin injection current: -15mA to 15mA unless Vdd > 4.9V, but that would not be the case when the connector is reversed and the protection diode is the only source of power to Vdd (that was not specified for classic AVRs, , but it's been said that the design maximum was +/- 1mA). The Dx-series is even better. Those spec clamp current of maximum +/- 20mA (yah, they changed the name they gave it between the two series)

**Reply by jvasileff** [reply]
I like the idea of using the FTDI pinout, but I'm trying to use 5 pins, not 6, to keep the size small. I'm sending the design off today - I suppose I'll keep my original pinout for the initial test board and adjust as necessary for the next one.

The board will switch UPDI vs RX,TX based on RTS. I'll share the design here in a couple weeks after I find out if it works!

**Reply by jvasileff** [reply]
> based on RTS

Ugh, it appears the current CH340 driver on macOS don't correctly set RTS. You have to first clear it before setting it, which terminal programs of course don't do.

In other words, this works:

```c
int fd = open("/dev/cu.usbserial-10", O_RDWR | O_NOCTTY);
int RTS_flag = TIOCM_RTS;
ioctl(fd, TIOCMBIC, &RTS_flag); // Clear RTS pin
ioctl(fd, TIOCMBIS, &RTS_flag); // Set RTS pin
```

but this doesn't:

```c
int fd = open("/dev/cu.usbserial-10", O_RDWR | O_NOCTTY);
int RTS_flag = TIOCM_RTS;
ioctl(fd, TIOCMBIS,&RTS_flag); // Set RTS pin
```

**Reply by SpenceKonde** [reply]
Yaaarrrgghhhh! Okay well, there goes my approach too. I suppose DTR is broken in the same way? if DTR works normally you just need to use a 340B or C (either works, both are SOP16 - but neither requires a the baggage of a crystal (naturally their baud rates have a bit more error from the inevitable inaccuracy of an internal RC instead of external crystal, but who wants to use

**Reply by jvasileff** [reply]
Yes, the DTR pin of the CH340C on the back of my discount Arduino Nano behaves the same way.

Interestingly, *either* RTS or DTR can be cleared to fix whatever errant initial state there is, to make all subsequent sets work. But that doesn't really help us.

**Reply by SpenceKonde** [reply]
One thought - try to scare up a CH340B, which has an RST line for an external reset pin, active low. Would be worth seeing if pulsing that low would result in behavior being corrected or not. Probably not, but if it did, one could come up with ways to fire that.

**Comment by jvasileff** [comment]
So I wound up building the auto-switching SerialUPDI board, and it works perfectly at 921600 baud.

Unfortunately, I haven't found a workaround for the macOS bug asserting RTS with Apple's driver. I tried the vendor's [driver](http://www.wch.cn/downloads/CH341SER_MAC_ZIP.html), and it does assert RTS correctly, but for some reason is painfully slow when programming at all baud rates.

![S08A7676](https://user-images.githubusercontent.com/236050/151421298-45960ba6-9479-44f8-96d1-33167d8a3eeb.jpeg)

**Comment by SpenceKonde** [comment]
Got a scope trace of the UPDI line during the slow programming?

**Reply by jvasileff** [reply]
I'll have to guess what will be helpful - if there's some other area you'd like to see, let me know.

A video of the terminal session to get an idea of what I'm seeing:

https://user-images.githubusercontent.com/236050/151443663-71482e6f-7b9c-4ea8-81da-102f8e160ba3.mov

First 70ms using *internal macOS driver*:

![macos-driver-first-70ms](https://user-images.githubusercontent.com/236050/151443724-464b3106-8e81-4971-8a90-28e30a1c857e.png)

First 70ms using vendor driver:

![vendor-driver-first-70ms](https://user-images.githubusercontent.com/236050/151443773-a3eed9ff-2dc5-4dae-afea-a38ec23265af.png)

Second grouping using vendor driver:

![vendor-driver-second-grouping](https://user-images.githubusercontent.com/236050/151443845-bbd1c5d1-879b-4636-bb97-494a5ea5cb28.png)
